### PR TITLE
Enable output of smoke and dust variables from RRFS

### LINKED ISSUE #575: Add smoke / dust output for RRFS
RRFS baseline E contains smoke and dust; we need to add this capability to UPP.  Code is working in GSL's RRFS.

### PR DESCRIPTION
**Pull request by EricJames-NOAA**
This PR enables output of smoke and dust related variables in RRFS, which is already used in GSL's RRFS-B and at least planned for EMC's RRFS-A.  

The code has been tested for GSL's RRFS-B in retrospective mode on Jet.

The regression tests pass except for RAP/HRRR and FV3R:

(1) For RAP/HRRR, it seems the authoritative repository does not have the UPP bug fix put into the realtime systems on 21 Dec 2021: https://www.weather.gov/media/notification/pdf2/scn21-86_rap_and_hrr_smoke_units_change_aab.pdf
This bug fix is needed to make the units consistent with labeling (kg/m3) on the smoke variable for RAP/HRRR.  The bug fix is included in this PR.

(2) For FV3R, a new baseline is needed since the inclusion of smoke and dust in UPP will affect the visibility diagnostic.  A change in the surface visibility field is the only difference in the current FV3R regression test.

### CODE REVIEW

File: sorc/ncep_post.fd/INITPOST_NETCDF.f [hunk 1]
```diff
@@ -1028,6 +1033,18 @@ SUBROUTINE INITPOST_NETCDF(ncid2d,ncid3d)
...
+       call read_netcdf_3d_para(ncid3d,im,jm,ista,ista_2l,iend,iend_2u,jsta,jsta_2l,jend,jend_2u, &
+       spval,recname(15),smoke(ista_2l,jsta_2l,1,1),lm)
+       call read_netcdf_3d_para(ncid3d,im,jm,ista,ista_2l,iend,iend_2u,jsta,jsta_2l,jend,jend_2u, &
+       spval,recname(16),fv3dust(ista_2l,jsta_2l,1,1),lm)
+       call read_netcdf_3d_para(ncid2d,im,jm,ista,ista_2l,iend,iend_2u,jsta,jsta_2l,jend,jend_2u, &
+       spval,recname(17),extsmoke(ista_2l,jsta_2l,1),lm)
+       call read_netcdf_3d_para(ncid2d,im,jm,ista,ista_2l,iend,iend_2u,jsta,jsta_2l,jend,jend_2u, &
+       spval,recname(18),extdust(ista_2l,jsta_2l,1),lm)
+       call read_netcdf_2d_para(ncid2d,ista,ista_2l,iend,iend_2u,jsta,jsta_2l,jend,jend_2u,       &
+       spval,recname(19),ebb(ista_2l,jsta_2l))
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/INITPOST_NETCDF.f]: @EricJames-NOAA Could you move reading ebb and hwp after line 1610? They should be in the block of reading variables from FV3 phy files. Please use the consistent format e.g. 
      VarName='maxvort02'
      call read_netcdf_2d_para(ncid3d,ista,ista_2l,iend,iend_2u,jsta,jsta_2l,jend,jend_2u, &
      spval,VarName,rel_vort_max(ista_2l,jsta_2l))

File: sorc/ncep_post.fd/INITPOST_NETCDF.f [hunk 2]
```diff
@@ -1028,6 +1033,18 @@ SUBROUTINE INITPOST_NETCDF(ncid2d,ncid3d)
        spval,recname(13),qqs(ista_2l,jsta_2l,1),lm)
        call read_netcdf_3d_para(ncid3d,im,jm,ista,ista_2l,iend,iend_2u,jsta,jsta_2l,jend,jend_2u, &
        spval,recname(14),qqg(ista_2l,jsta_2l,1),lm)
+       call read_netcdf_3d_para(ncid3d,im,jm,ista,ista_2l,iend,iend_2u,jsta,jsta_2l,jend,jend_2u, &
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/INITPOST_NETCDF.f:1042]: @EricJames-NOAA Can you add an  IF statement of modelname as FV3R for reading recname 15-18?

File: sorc/ncep_post.fd/INITPOST_NETCDF.f (hunk 1 above)
Comment (EricJames-NOAA) [reply sorc/ncep_post.fd/INITPOST_NETCDF.f]: @WenMeng-NOAA thanks for the suggestion, I made this change

File: sorc/ncep_post.fd/INITPOST_NETCDF.f (hunk 2 above)
Comment (EricJames-NOAA) [reply sorc/ncep_post.fd/INITPOST_NETCDF.f:1042]: @WenMeng-NOAA thanks, I added this IF statement.

File: sorc/ncep_post.fd/INITPOST_NETCDF.f [hunk 3]
```diff
@@ -2978,6 +3003,27 @@ SUBROUTINE INITPOST_NETCDF(ncid2d,ncid3d)
...
       END IF
+!
+! E. James - 27 Sep 2022: this is for RRFS, adding smoke and dust
+! extinction; it needs to be after ZINT is defined.
+!
+      if (modelname == 'FV3R') then
+       do l = 1, lm
+        do j = jsta_2l, jend_2u
+         do i = ista_2l, iend_2u
+            taod5503d ( i, j, l) = extsmoke ( i, j, l ) + extdust ( i, j, l )
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/INITPOST_NETCDF.f:3020]: @EricJames-NOAA If variables extsmoke and extdust are not used other in  routines, you might locally define, allocate and deallocate them in this routine for saving memory usage.

Comment (EricJames-NOAA) [reply sorc/ncep_post.fd/INITPOST_NETCDF.f:3020]: Good idea...I have implemented that.  Let me know if you think any of the allocation / deallocation should be moved elsewhere within INITPOST_NETCDF.f

File: sorc/ncep_post.fd/DEALLOCATE.f [hunk 4]
```diff
@@ -197,9 +197,9 @@ SUBROUTINE DE_ALLOCATE
       deallocate(swddnic)
       deallocate(swddifc)
       deallocate(swupt)
-      deallocate(int_smoke)
       deallocate(mean_frp)
-      deallocate(int_aod)
+      deallocate(ebb)
+      deallocate(hwp)
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/DEALLOCATE.f:202]: @EricJames-NOAA Could you also add a deallocate statement for fv3dust?

File: sorc/ncep_post.fd/INITPOST_NETCDF.f [hunk 5]
```diff
@@ -2978,6 +3008,29 @@ SUBROUTINE INITPOST_NETCDF(ncid2d,ncid3d)
...
+            dz = ZINT( i, j, l ) - ZINT( i, j, l+1 )
+            aextc55 ( i, j, l ) = taod5503d ( i, j, l ) / dz
+         if(i==im/2.and.j==(jsta+jend)/2)print*,'sample taod5503d= ',   &
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/INITPOST_NETCDF.f]: @EricJames-NOAA Could you also add a threshold "debugprint" for these three print statements? 
e.g. if(debugprint.and.i==im/2.and.j==(jsta+jend)/2)print*,

**Comment by WenMeng-NOAA** [comment]
@EricJames-NOAA There is a bug introduced from your PR #582. Could you add the following fixes in this PR in INITPOST_NETCDF.f as:
Change
   ```
      do j=jsta,jend
        do i=1,im
          smstav(i,j) = buf(i,j)
        enddo
      enddo
   ```
into 
  ```
      do j=jsta,jend
        do i=ista,iend
          smstav(i,j) = buf(i,j)
        enddo
      enddo
  ```

**Comment by EricJames-NOAA** [comment]
> @EricJames-NOAA There is a bug introduced from your PR #582. Could you add the following fixes in this PR in INITPOST_NETCDF.f as: Change
> 
> ```
>    do j=jsta,jend
>      do i=1,im
>        smstav(i,j) = buf(i,j)
>      enddo
>    enddo
> ```
> 
> into
> 
> ```
>     do j=jsta,jend
>       do i=ista,iend
>         smstav(i,j) = buf(i,j)
>       enddo
>     enddo
> ```

@WenMeng-NOAA very sorry for the bug!  I will put in this fix shortly.  Thanks for your other good suggestions too.

File: sorc/ncep_post.fd/MDL2P.f [hunk 6]
```diff
@@ -148,6 +149,15 @@ SUBROUTINE MDL2P(iostatusD3D)
           enddo
         enddo
       enddo
+      if (.not. allocated(fv3dustsl)) allocate(fv3dustsl(im,jm,nbin_sm))
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/MDL2P.f:152]: @EricJames-NOAA Could you add deallocate statement for fv3dustsl after line 3847?

**Comment by WenMeng-NOAA** [comment]
@EricJames-NOAA There is another bug introduced from my PR #589. Would you mind adding the fix in this PR?
In scripts/exgfs_atmos_nceppost.sh:
change
   rdaod=.true.
into
   rdaod=.true.,

Thanks!

File: parm/post_avblflds.xml [hunk 7]
```diff
@@ -6758,35 +6758,39 @@
     <param>
        <post_avblfldidx>736</post_avblfldidx>
        <shortname>SMOKE_ON_ENTIRE_ATMOS_SINGLE_LYR</shortname>
-       <pdstmpl>tmpl4_0</pdstmpl>
+       <pdstmpl>tmpl4_48</pdstmpl>
```
Comment (WenMeng-NOAA) [inline parm/post_avblflds.xml:6761]: @EricJames-NOAA Could you also add these tmpl4_48 updates in post_avblflds_raphrrr.xml which is used for rap/hrrr control files. After updating, you may prompt "make" under parm/ to generate new flat text control files for rap/hrrr and rtma. Please also commit these new control files.

File: sorc/ncep_post.fd/CLDRAD.f [hunk 8]
```diff
@@ -3831,49 +3851,83 @@ SUBROUTINE CLDRAD
       ENDIF
 
 ! Instantaneous clear-sky downwelling LW at the surface
-      IF (IGET(744)>0) THEN
+      IF (IGET(764)>0) THEN
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/CLDRAD.f:3854]: @EricJames-NOAA Thanks for catching the issue of instant clear-sky downwelling/upwelling long-wave flux having no UPP IDs. Could you add these two definitions in post_avblflds.xml as:
```
    <param>
       <post_avblfldidx>764</post_avblfldidx>
       <shortname>CSDLF_ON_SURFACE</shortname>
       <pname>CSDLF</pname>
       <fixed_sfc1_type>surface</fixed_sfc1_type>
       <scale>3.0</scale>
    </param>
  
       <param>
       <post_avblfldidx>765</post_avblfldidx>
       <shortname>CSULF_ON_SURFACE</shortname>
       <pname>CSULF</pname>
       <fixed_sfc1_type>surface</fixed_sfc1_type>
       <scale>3.0</scale>
    </param>
``` 
We might use them in the future. Thanks!

**Comment by WenMeng-NOAA** [comment]
@EricJames-NOAA I conducted the UPP standalone test with RRFS data you provided. The results are at /scratch1/NCEPDEV/stmp2/Wen.Meng/rrfs_2022102713. Can you verify if my test is consistent with the test from your side?

File: sorc/ncep_post.fd/INITPOST_NETCDF.f [hunk 9]
```diff
@@ -2978,6 +3008,29 @@ SUBROUTINE INITPOST_NETCDF(ncid2d,ncid3d)
...
+! E. James - 27 Sep 2022: this is for RRFS, adding smoke and dust
+! extinction; it needs to be after ZINT is defined.
+!
+      if (modelname == 'FV3R') then
+       do l = 1, lm
+        do j = jsta_2l, jend_2u
+         do i = ista_2l, iend_2u
+            taod5503d ( i, j, l) = extsmoke ( i, j, l ) + extdust ( i, j, l )
+            dz = ZINT( i, j, l ) - ZINT( i, j, l+1 )
+            aextc55 ( i, j, l ) = taod5503d ( i, j, l ) / dz
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/INITPOST_NETCDF.f:3022]: @EricJames-NOAA Could you add thresholds with extsmoke and extdust in valid values in this aextc55 calculation loop? e.g.
```
if(extsmoke(i,j,l) <spval .and. extdust(i,j,l)<spval) then
  ...
```

**Comment by EricJames-NOAA** [comment]
> 

@WenMeng-NOAA your test results look consistent with mine.  Thanks for your test!

**Comment by WenMeng-NOAA** [comment]
@EricJames-NOAA With your latest commit, the changed visibility results in fv3r were reverted. The current model outputs for fv3r do not include extdust and extsmoke.

**Review by WenMeng-NOAA — APPROVED** [review]

**Comment by WenMeng-NOAA** [comment]
The UPP RT tests were completed on WCOSS2, Hera and Orion. The new base lines are needed for rap/hrrr, 3drtma with this PR.

**Pull request merged**